app/services/authentication/authenticator.rb

```diff
@@ -0,0 +1,127 @@
+module Authentication
+  # Authenticator will perform one of these tree operations:
+  # 1. create a new user and match it to its authentication identity
+  # 1. update an existing user and align it to its authentication identity
+  # 1. return the current user if a user is given (already logged in scenario)
+  class Authenticator
+    # auth_payload is the payload schema, see https://github.com/omniauth/omniauth/wiki/Auth-Hash-Schema
+    def initialize(auth_payload, current_user: nil, cta_variant: nil)
+      @provider = load_authentication_provider(auth_payload)
+
+      @current_user = current_user
+      @cta_variant = cta_variant
+    end
+
+    def self.call(*args)
+      new(*args).call
+    end
+
+    def call
+      identity = Identity.build_from_omniauth(provider)
+
+      return current_user if current_user_identity_exists?
+
+      user = proper_user(identity)
+      user = if user.nil?
+               build_user
+             else
+               update_user(user)
+             end
+
+      save_identity(identity, user)
+
+      user.skip_confirmation!
+
+      flag_spam_user(user) if account_less_than_a_week_old?(user, identity)
+
+      user.save!
+      user
+    end
+
+    private
+
+    attr_reader :provider, :current_user, :cta_variant
+
+    # Loads the proper authentication provider from the available ones
+    # TODO: [thepracticaldev/oss] raise exception if provider is available but not enabled for this app
+    # TODO: [thepracticaldev/oss] add available providers, enabled providers
+    def load_authentication_provider(auth_payload)
+      provider_name = auth_payload.provider
+      provider_class = "Authentication::Providers::#{provider_name.titleize}".constantize
+      provider_class.new(auth_payload)
+    rescue NameError => e
+      raise ::Authentication::Errors::ProviderNotFound, e
```

Comment:
some magic here. we load the provider class at runtime, in the future we'll also check if the provider is enabled, and not just available

Reply:
I think the `Authenticator` should know less. Maybe something like the following instead?

```ruby
provider_class = Authentication::Providers.get!(provider_name)
provider_class.new(auth_payload)
```

We can then raise whatever error we want there (e.g. `::Authentication::Errors::ProviderNotFound` and `::Authentication::Errors::ProviderNotEnabled`) and let them bubble up.

Reply:
I like the suggestion, going to change it in this PR